### ISSUE TITLE
particle spawner (vizlib)

### ISSUE THREAD

**Issue by SwissalpS** [pandorabox-io/in-game#317]
I love vizlib. It's easy to implement cool markers now.

There is something I need to get out there though:
it's heavy on GPU

Hitting a quarry a dozen times in rapid succession, blocks my client. FPS falls to 1 and movement input is ignored until those markers are gone.

Potentially this is a grief vector. Player A with a powerful machine can freeze all players with weaker hardware that are in the area.

Please add throtles to mods that you have added vizlib or other particle spawner markers.
Basically make it less easy for malicious players.

Thanks :D

**Comment by S-S-X** [comment]
Isn't vizlib already limiting particles in these cases so that particles are only sent to player who activates it?
And node/entity/whatever should limit by themselves in case where visualization data is actually sent to all players?

Sure it might be useful in some cases if vizlib could provide simple shortcut for this but should not be automatically limited because that would either limit possible vizlib use cases a lot or would require a lot of potentially unnecessary checks for different situations.

**Comment by S-S-X** [comment]
For vizlib rate limiting shortcut and future API development suggestion, one possibility by providing wide spread de facto standard API that would allow rate limiters and a lot more (even direct complete particle control if MT has it one day) without increasing API complexity at all, being similar to MT core APIs and additionally not necessarily increasing resource usage at all.

I'd suggest `vizlib.draw_*` could return id and provide another API method to get list of particle id's at any give position, something like `vizlib.get_visualizations(id|pos|corner_pos|area)` and for rate limiting return at least visualization lifetimes.

This would be pretty much standard way to provide easy to use extensible way to add more functionality in future and currently could allow checking if any given position currently has active particles for player in question.
Additionally API like that would be easily extensible for any particle parameters, particle control already available and particle control possibly available in future.
Also API like that wont increase regular API method complexity at all for API users.

If you feel you do not understand what I mean just think how for example MT sound API or HUD API works for publicly accessible user facing parts and under the hood.

**Comment by SwissalpS** [comment]
I did not mean vizlib to throttle but the mods that use it or particles by themselves.  
Then again, yes why not build it into vizlib.

Yes vizlib offers to only show to single player, not all mods use that.

There is a PR in the works to add 3d lines to minetest. That should be less intensive on hardware than moving particles acting like lines.

**Comment by OgelGames** [comment]
One possible solution could be to block the creation of identical particle spawners.

But really each node creating the shapes should handle the rate limiting, it could be as simple as storing the expiry time for the shape, and checking if it has expired before creating another one:

```lua
on_punch = function(pos, _, player)
	if not player or player:get_wielded_item():get_name() ~= "" then
		return
	end
	local meta = minetest.get_meta(pos)
	if meta:get_int("shape_expiry") > os.time() then
		return
	end
	local shape = vizlib.draw_cube(pos, meta:get_int("radius") + 0.5, {player = player})
	meta:set_int("shape_expiry", shape.expiry)
end
```

Storing the shapes by ID could be useful for other reasons though, so maybe an API for that could be added too.

Now if only github would let me transfer this issue to the `vizlib` repo... 😖
